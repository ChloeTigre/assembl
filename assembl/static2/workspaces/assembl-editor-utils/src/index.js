// @flow
import { EditorState, Modifier, RichUtils } from 'draft-js';
import DraftJSPluginsUtils from 'draft-js-plugins-utils';

import { getSelectedBlockKey } from './blocks';
import * as constants from './constants';
import { getEntityRange } from './entities';
import { createSelectionState } from './selection';

type LinkData = {
  target: ?string,
  text: ?string,
  title: ?string,
  url: string
};

const { collapseToEnd, getCurrentEntity, getCurrentEntityKey, hasEntity, removeLinkAtSelection } = DraftJSPluginsUtils;

const { ENTITY_TYPES, ENTITY_MUTABILITY } = constants;

export default {
  collapseToEnd: collapseToEnd,

  createLinkAtSelection: (editorState: EditorState, data: LinkData): EditorState => {
    const { text } = data;
    const contentState = editorState.getCurrentContent().createEntity(ENTITY_TYPES.link, ENTITY_MUTABILITY.mutable, data);
    const entityKey = contentState.getLastCreatedEntityKey();
    let withLink;
    if (text) {
      const newContentState = Modifier.replaceText(contentState, editorState.getSelection(), text, undefined, entityKey);
      withLink = EditorState.createWithContent(newContentState);
    } else {
      withLink = RichUtils.toggleLink(editorState, editorState.getSelection(), entityKey);
    }
    return EditorState.forceSelection(withLink, editorState.getSelection());
  },

  replaceLinkAtCursor: (editorState: EditorState, data: LinkData): EditorState => {
    const { text } = data;
    const contentState = editorState.getCurrentContent();
    const blockKey = getSelectedBlockKey(editorState);
    const replacedEntityKey = getCurrentEntityKey(editorState);
    if (replacedEntityKey) {
      const entityRange = getEntityRange(editorState, replacedEntityKey);
      if (text && entityRange) {
        const { start, end } = entityRange;
        const replacedSelection = createSelectionState(blockKey, start, end);
        const newContentState = Modifier.replaceText(contentState, replacedSelection, text, null, replacedEntityKey);
        newContentState.mergeEntityData(replacedEntityKey, data);
        return EditorState.createWithContent(newContentState);
      }
    }

    return editorState;
  },

  getCurrentEntityKey: getCurrentEntityKey,

  getCurrentEntity: getCurrentEntity,

  hasEntity: hasEntity,

  removeLinkAtSelection: removeLinkAtSelection
};

export { constants, createSelectionState };