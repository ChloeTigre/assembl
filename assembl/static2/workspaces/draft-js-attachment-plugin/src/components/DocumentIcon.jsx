// @flow
import * as React from 'react';
import { ContentState } from 'draft-js';
import type { ContentBlock } from 'draft-js';

import { constants } from 'assembl-editor-utils';

import getDocumentIconPath from '../utils/getDocumentIconPath';
import getFileExtension from '../utils/getFileExtension';

type Props = {
  block: ContentBlock,
  contentState: ContentState
};

const DocumentIcon = ({ block, contentState }: Props) => {
  const entityKey = block.getEntityAt(0);
  const entity = contentState.getEntity(entityKey);
  const data = entity.getData();
  const type = entity.getType();
  if (type === constants.ENTITY_TYPES.document) {
    const extension = getFileExtension(data.title);
    const iconPath = getDocumentIconPath(extension);
    return (
      <div className="atomic-block" data-blocktype="atomic" title={data.title}>
        <img className="attachment-icon" src={iconPath} alt={extension} />
      </div>
    );
  }

  return <div />;
};

export default DocumentIcon;