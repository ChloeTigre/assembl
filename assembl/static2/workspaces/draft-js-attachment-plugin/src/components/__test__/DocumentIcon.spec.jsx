// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { AtomicBlockUtils, ContentState, EditorState } from 'draft-js';

import { constants } from 'assembl-editor-utils';

import DocumentIcon from '../DocumentIcon';

const { ENTITY_MUTABILITY, ENTITY_TYPES } = constants;

describe('DocumentIcon component', () => {
  it('should render a document icon', () => {
    let contentState = ContentState.createFromText('');
    // $FlowFixMe DraftEntityType is too restrictive in DraftJS (see https://github.com/facebook/draft-js/issues/868 )
    contentState = contentState.createEntity(ENTITY_TYPES.document, ENTITY_MUTABILITY.immutable, {
      externalUrl: 'my-file.pdf',
      mimeType: 'application/pdf',
      title: 'My file'
    });
    const entityKey = contentState.getLastCreatedEntityKey();
    let editorState = EditorState.createWithContent(contentState);
    editorState = AtomicBlockUtils.insertAtomicBlock(editorState, entityKey, ' ');
    contentState = editorState.getCurrentContent();
    const blocks = contentState.getBlocksAsArray();
    const atomicBlock = blocks[1];

    const props = {
      block: atomicBlock,
      contentState: contentState
    };
    const component = renderer.create(<DocumentIcon {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});