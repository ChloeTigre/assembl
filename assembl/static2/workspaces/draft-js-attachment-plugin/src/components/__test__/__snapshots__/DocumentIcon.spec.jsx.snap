// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentIcon component should render a document icon 1`] = `
<div
  className="atomic-block"
  data-blocktype="atomic"
  title="My file"
>
  <img
    alt="unknown"
    className="attachment-icon"
    src="/static2/img/icons/black/doc.svg"
  />
</div>
`;
